src/ObjCRuntime/Runtime.cs

```diff
@@ -1304,19 +1304,46 @@ static void MissingCtor (IntPtr ptr, IntPtr klass, Type type, MissingCtorResolut
 			return ConstructNSObject<T> (ptr, typeof (T), MissingCtorResolution.ThrowConstructor1NotFound);
 		}
 
-		// The generic argument T is only used to cast the return value.
 		// The 'selector' and 'method' arguments are only used in error messages.
-		static T? ConstructNSObject<T> (IntPtr ptr, Type type, MissingCtorResolution missingCtorResolution) where T : class, INativeObject
+		static T? ConstructNSObject<T> (IntPtr ptr, Type type, MissingCtorResolution missingCtorResolution) where T : NSObject
 		{
 			return ConstructNSObject<T> (ptr, type, missingCtorResolution, IntPtr.Zero, default (RuntimeMethodHandle));
 		}
 
-		// The generic argument T is only used to cast the return value.
 		// The 'selector' and 'method' arguments are only used in error messages.
-		static T? ConstructNSObject<T> (IntPtr ptr, Type type, MissingCtorResolution missingCtorResolution, IntPtr sel, RuntimeMethodHandle method_handle) where T : class, INativeObject
+		static T? ConstructNSObject<T> (IntPtr ptr, Type type, MissingCtorResolution missingCtorResolution, IntPtr sel, RuntimeMethodHandle method_handle)
+			where T : NSObject
+#if NET
+				, INSObjectFactory
+#endif
 		{
 			if (type is null)
 				throw new ArgumentNullException (nameof (type));
+#if NET
+			if (Runtime.IsManagedStaticRegistrar) {
+				T? instance = default;
+				var nativeHandle = new NativeHandle (ptr);
+
+				if (typeof (T) != typeof (NSObject)
+					&& (typeof (T) == type || typeof (T).IsGenericType)
+					&& !(typeof (T).IsInterface || typeof (T).IsAbstract)) {
+					instance = ConstructNSObjectViaFactoryMethod (nativeHandle);
+				}
+
+				// If we couldn't create an instance of T through the factory method, we'll use the lookup table.
+				// This table can't instantiate generic types though.
+				if (!type.IsGenericType) {
+					instance ??= RegistrarHelper.ConstructNSObject<T> (type, nativeHandle);
+				}
+
+				if (instance is null) {
+					MissingCtor (ptr, IntPtr.Zero, type, missingCtorResolution, sel, method_handle);
```

Comment:
We might need a different error message, because it would be rather confusing for customers if we said we couldn't find the constructor and it ended up actually being there.